pkg/scale/decode.go

```diff
@@ -523,10 +524,68 @@ func (ds *decodeState) decodeUint(dstv reflect.Value) (err error) {
 	return
 }
 
+var errorU32OutOfRange = errors.New("uint32 out of range")
+
+// decodeCompactUint32 will decode compact unsigned 32-bit integer
+func (ds *decodeState) decodeCompactUint32(dstv reflect.Value) (err error) {
+	b, err := ds.ReadByte()
+	if err != nil {
+		return
+	}
+
+	in := dstv.Interface()
+	temp := reflect.New(reflect.TypeOf(in))
+	// check mode of encoding, stored at 2 least significant bits
+	mode := b & 3
+	switch {
+	case mode <= 2:
+		var val int64
+		val, err = ds.decodeSmallInt(b, mode)
+		if err != nil {
+			return
+		}
+		if mode == 1 {
+			if !(val > 0b0011_1111 && val <= 0b0011_1111_1111_1111) {
+				return errorU32OutOfRange
+			}
+		}
+		if mode == 2 {
+			if !(val > 0b0011_1111_1111_1111 && val <= math.MaxUint32>>2) {
+				return errorU32OutOfRange
+			}
+		}
+		temp.Elem().Set(reflect.ValueOf(val).Convert(reflect.TypeOf(in)))
+		dstv.Set(temp.Elem())
+	default:
+		// >4 byte mode
+		topSixBits := b >> 2
+		if mode == 3 && topSixBits != 0 {
+			return errorU32OutOfRange
+		}
+		byteLen := uint(topSixBits) + 4
+
+		buf := make([]byte, byteLen)
+		_, err = ds.Read(buf)
```

Comment:
Aren't we just reading 0 bytes from that buffer? 🤔 Or the ds.Read is badly named and should be `Write` instead??